### ISSUE TITLE
SEO: Rewrite Module

### ISSUE THREAD

**Issue by Flarescape** [OrchardCMS/OrchardCore#9027]
I know that there are some open issues out there, where other people asking for an extension to the autoroute module for path updates and redirects, but this does not cover actual SEO relevant redirects.

In O1 there was the so called Contrib.RewriteRules module, which worked with apache mod rewrite rules syntax.
I've just created a custom rewrite module with the same functionality for OC, but with much less code.
It's very helpfull that we now have "app.UseRewriter", which supports "RewriteOptions" and automatically parses apache mod rewrite rule syntax, so no need for a custom parser, but i don't think that we even need to use the apache rules, the build in syntax is much better.

So i think, that a rewrite rules module in the SEO section of the admin menu would be great.
This could be easily achived by creating a content type with a list part of content type "rule".
I think, a single roule needs a "From" and a "To" field and some options for the type of redirect, like permanent, etc.
And of course support for wildecards.

Maybe this could be a standard part of the already existing SEO module, because i would consider this as a prerequisite for the Website/CMS part of OC.

**Comment by jtkech** [comment]
> I've just created a custom rewrite module with the same functionality for OC, but with much less code.

Cool, are you open to create a PR based on your code? That would suggest a separate module / feature or as you think an additional feature of the SEO module

**Comment by deanmarcussen** [comment]
Yes, sounds great. We discussed it at a meeting either last week or the week before, if someone can find the video.

Probably a separate module. But open a pr please

**Comment by Piedone** [comment]
This would also be quite useful when migrating sites to Orchard and you want to keep old URLs working.

**Comment by xperiandri** [comment]
Any updates? We need redirects functionality.
I thought about implementation of them as the ability to add a list of URLs to the main URL of AutoRoute. Where you add a rules for content type in AutoRoute module and they will redirect to that content item.

**Comment by xperiandri** [comment]
@sebastienros can we implement that?
By adding a section on the SEO tab in a content item

**Comment by deanmarcussen** [comment]
Seperate Rewrite module. RewritePart for content items

plus would need to handle rules that are not content item related, or support for a decoupled razor page scenario.

Also discussed the need for a url tracker for tracking not found's. But that's a seperate feature.

Would also still be good to see what you have already done @Flarescape 

For some ideas
https://github.com/EtchUK/Etch.OrchardCore.SEO

**Comment by Piedone** [comment]
For now, you can use the [URL Rewriting Middleware](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/url-rewriting?view=aspnetcore-8.0) to add any redirects.

Autoroute redirects issue: https://github.com/OrchardCMS/OrchardCore/issues/8864.
